wafer/compare/admin.py

```diff
@@ -0,0 +1,155 @@
+from django.core.urlresolvers import reverse
+from django.utils.translation import ugettext as _
+from reversion.helpers import generate_patch_html
+from django.contrib.admin import SimpleListFilter
+from django.contrib.contenttypes.models import ContentType
+
+from markitup.fields import Markup
+
+
+class DateModifiedFilter(SimpleListFilter):
+    title = _('Last Modified')
+    parameter_name = 'moddate'
+
+    def lookups(self, request, model_admin):
+        return (
+            ('today', _('Today')),
+            ('yesterday', _('Since yesterday')),
+            ('7 days', _('In the last 7 days')),
+            ('30 days', _('In the last 30 days')),
+            )
+
+    def queryset(self, request, queryset):
+        date = None
+        if self.value() == 'today':
+            date = datetime.date.today()
+        elif self.value() == 'yesterday':
+            date = datetime.date.today() - datetime.timedelta(days=1)
+        elif self.value() == '7 days':
+            date = datetime.date.today() - datetime.timedelta(days=7)
+        elif self.value() == '30 days':
+            date = datetime.date.today() - datetime.timedelta(days=30)
+        if not date:
+            return queryset
+        content_type = ContentType.objects.get_for_model(queryset.model)
+        revisions = Version.objects.filter(content_type=content_type, revision__date_created__gte=date)
+        return queryset.filter(pk__in=[x.object_id for x in revisions])
+
+
+class CompareVersionAdmin(VersionAdmin):
+
+    compare_template = "admin/wafer.compare/compare.html"
+    compare_list_template = "admin/wafer.compare/compare_list.html"
+
+    # Add a compare button next to the History button.
+    change_form_template = "admin/wafer.compare/change_form.html"
+
+    def get_urls(self):
+         urls = super(CompareVersionAdmin, self).get_urls()
+         opts = self.model._meta
+         compare_urls = [
+               url("^([^/]+)/([^/]+)/compare/$", self.admin_site.admin_view(self.compare_view),
+                   name='%s_%s_compare' % (opts.app_label, opts.model_name)),
+               url("^([^/]+)/comparelist/$", self.admin_site.admin_view(self.comparelist_view),
+                   name='%s_%s_comparelist' % (opts.app_label, opts.model_name)),
+         ]
+         return compare_urls + urls
+
+    def compare_view(self, request, object_id, version_id, extra_context=None):
+        """Actually compare two versions."""
+        opts = self.model._meta
+        object_id = unquote(object_id)
+        # get_for_object's ordering means this is always the latest revision.
+        current = self.revision_manager.get_for_object_reference(self.model, object_id)[0]
+        # The reversion we want to compare to
+        revision = self.revision_manager.get_for_object_reference(self.model, object_id).filter(id=version_id)[0]
+        the_diff = []
+        dmp = diff_match_patch()
+
+        for field in current.field_dict:
```

Comment:
Extending the field list with those from the old version is probably sensible.

All the diffs are sorted (line 115) before being passed to the template, so the ordering should always be consistent.